SU2_CFD/src/numerics_structure.cpp

```diff
@@ -2810,6 +2920,285 @@ void CNumerics::SetRoe_Dissipation(const su2double Dissipation_i,
 
 }
 
+void CNumerics::EigenDecomposition(su2double **A_ij, su2double **Eig_Vec, su2double *Eig_Val){
+  int iDim,jDim;
+  su2double *e = new su2double [3];
+  for (iDim= 0; iDim< 3; iDim++){
+    e[iDim] = 0;
+    for (jDim = 0; jDim < 3; jDim++){
+      Eig_Vec[iDim][jDim] = A_ij[iDim][jDim];
+    }
+  }
+  tred2(Eig_Vec, Eig_Val, e);
+  tql2(Eig_Vec, Eig_Val, e);
+
+  delete [] e;
+}
+
+void CNumerics::EigenRecomposition(su2double **A_ij, su2double **Eig_Vec, su2double *Eig_Val){
+  unsigned short i,j,k;
+  su2double **tmp = new su2double* [3];
+
+  for (i= 0; i< 3; i++){
+    tmp[i] = new su2double [3];
+  }
+
+  for (i= 0; i< 3; i++){
+    for (j = 0; j < 3; j++){
+      tmp[i][j] = 0.0;
+      for (k = 0; k < 3; k++){
+        tmp[i][j] += Eig_Vec[i][k] * Eig_Val[k] * delta3[k][j];
+      }
+    }
+  }
+
+  for (i= 0; i< 3; i++){
+    for (j = 0; j < 3; j++){
+      A_ij[i][j] = 0.0;
+      for (k = 0; k < 3; k++){
+        A_ij[i][j] += tmp[i][k] * Eig_Vec[j][k];
+      }
+    }
+  }
+
+  for (i = 0; i < 3; i++){
+    delete [] tmp[i];
+  }
+  delete [] tmp;
+}
+
+void CNumerics::tred2(su2double **V, su2double *d, su2double *e) {
+
+    unsigned short i,j,k;
+    /*  This is derived from the Algol procedures tred2 by        */
```

Comment:
Attribution here is good to have, but do they give specific guidance about the usage of this code (is it open)? Is there a specific license or statement we need to include? Or did you write it yourself with only inspiration from their work?

Reply:
This was essentially converted from fortran code (https://netlib.sandia.gov/eispack/tql2.f) to be used with c++. I attributed it according to how it was attributed in the fortran code. Not sure what exactly is the policy on that. Since the code was adapted, not exactly copied. 

This is from EISPACK which is now superseded by LAPACK

Reply:
I couldn't find the license explicitly for EISPACK. It seems like they might not have thought about it when they wrote those routines and put them online. But they have the modified BSD license on the LAPACK website: http://www.netlib.org/lapack/LICENSE.txt 

As for the modified C++ routines, they would require some changes as well as it has an unnecessary output that requires additional computations. In the end, it would look identical to what is in the code currently, perhaps with a couple of more helper variables. 

Since the code would end up being the same, I would prefer attributing to the original EISPACK implementation. Does that mean that I should provide the link to the license file in the code?

Reply:
Yep. We should provide the original license, or some written permission to use the code from the author/copyright holder. If you can find the original license, we can embed it in the comments in this file, or we can ship the file with the rest of the code.

Reply:
So I have repeatedly emailed the LAPACK email list about the EISPACK license but haven't gotten a reply back from them. So I am not sure how to proceed. The only license I have found is still for LAPACK. 

I also found this thread: http://icl.cs.utk.edu/lapack-forum/archives/lapack/msg00202.html 

It seems to say that there is no licensing restrictions on EISPACK code. We could reference it in the way it is referenced in the c++ version of the code by jburkardt that you linked in the previously, which I have copied here: 

//  Author:
//
//    Original FORTRAN77 version by Smith, Boyle, Dongarra, Garbow, Ikebe,
//    Klema, Moler.
//    C version by John Burkardt.
//
//  Reference:
//
//    Martin, Reinsch, Wilkinson,
//    TRED2,
//    Numerische Mathematik,
//    Volume 11, pages 181-195, 1968.
//
//    James Wilkinson, Christian Reinsch,
//    Handbook for Automatic Computation,
//    Volume II, Linear Algebra, Part 2,
//    Springer, 1971,
//    ISBN: 0387054146,
//    LC: QA251.W67.
//
//    Brian Smith, James Boyle, Jack Dongarra, Burton Garbow,
//    Yasuhiko Ikebe, Virginia Klema, Cleve Moler,
//    Matrix Eigensystem Routines, EISPACK Guide,
//    Lecture Notes in Computer Science, Volume 6,
//    Springer Verlag, 1976,
//    ISBN13: 978-3540075462,
//    LC: QA193.M37.

Reply:
Ok, please add a statement similar to the reference that you have pasted here.. at least there is some precedent for that from the linked website. If you hear anything more from email, we can update accordingly.

Reply:
Maybe one last comment: you could also try reaching out personally to one of the original authors to ask for explicit permission and document that.